tox/config.py

```diff
@@ -913,8 +913,20 @@ def _split_env(env):
 
 
 def _split_factor_expr(expr):
+    def split_single(e):
+        raw = e.split('-')
+        included = {x for x in raw if x[:1] != '!'}
+        excluded = {x[1:] for x in raw if x[:1] == '!'}
+        return included, excluded
+
+    partial_envs = _expand_envstr(expr)
+    return [split_single(e) for e in partial_envs]
+
+
+def _split_factor_expr_all(expr):
     partial_envs = _expand_envstr(expr)
-    return [set(e.split('-')) for e in partial_envs]
+    return [{x[1:] if x[:1] == '!' else x for x in e.split('-')}
```

Comment:
I feel like we should have a ``is_negated_factor`` function to avoid all this repetition of this check

Reply:
the variable x is a bit too cryptic, e.g. the ``partial_envs`` hints that ``e`` will be a ``partial env``, but once I iterate on that what that contains?